File: app/src/main/java/me/blog/korn123/commons/utils/CommonUtils.java
package me.blog.korn123.commons.utils;

import android.content.ContentResolver;
import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.util.TypedValue;

/**
 * Created by CHO HANJOONG on 2017-03-16.
 */

public class CommonUtils {

    public static int getBuildVersion() {
        return android.os.Build.VERSION.SDK_INT;
    }

    public static String loadStringPreference(Context context, String key, String defaultValue) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getString(key, defaultValue);
    }

    public static void saveStringPreference(Context context, String key, String value) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        SharedPreferences.Editor edit = preferences.edit();
        edit.putString(key, value);
        edit.commit();
    }

    public static long loadLongPreference(Context context, String key, int defaultValue) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getLong(key, defaultValue);
    }

    public static void saveLongPreference(Context context, String key, long value) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        SharedPreferences.Editor edit = preferences.edit();
        edit.putLong(key, value);
        edit.commit();
    }

    public static float loadFloatPreference(Context context, String key, int defaultValue) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getFloat(key, defaultValue);
    }

    public static void saveFloatPreference(Context context, String key, float value) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        SharedPreferences.Editor edit = preferences.edit();
        edit.putFloat(key, value);
        edit.commit();
    }

    public static int loadIntPreference(Context context, String key, int defaultValue) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getInt(key, defaultValue);
    }

    public static void saveIntPreference(Context context, String key, int value) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        SharedPreferences.Editor edit = preferences.edit();
        edit.putInt(key, value);
        edit.commit();
    }

    public static boolean loadBooleanPreference(Context context, String key) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        boolean result = preferences.getBoolean(key, false);
        return result;
    }

    public static void saveBooleanPreference(Context context, String key, boolean isEnable) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        SharedPreferences.Editor edit = preferences.edit();
        edit.putBoolean(key, isEnable);
        edit.commit();
    }

    public static int dpToPixel(Context context, int dp) {
        return dpToPixel(context, dp, 0);
    }

    public static int dpToPixel(Context context, int dp, int policy) {
        float px = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, context.getResources().getDisplayMetrics());
        int pixel = 0;
        switch (policy) {
            case 0:
                pixel = (int) px;
                break;
            case 1:
                pixel = Math.round(px);
                break;
        }
        return pixel;
    }

    public static String uriToPath(ContentResolver contentResolver, Uri uri) {
        String path = null;
        Cursor cursor = contentResolver.query(uri, null, null, null, null);
        cursor.moveToNext();
        int columnIndex = cursor.getColumnIndex("_data");
        if (columnIndex > 0) {
            path = cursor.getString(columnIndex);
        } else {
            path = uri.toString();
        }
        cursor.close();
        return path;
    }
}

File: app/src/main/java/me/blog/korn123/commons/utils/CommonUtils.kt
package me.blog.korn123.commons.utils

import android.app.Activity
import android.content.ContentResolver
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Point
import android.net.Uri
import android.util.TypedValue
import id.zelory.compressor.Compressor
import me.blog.korn123.easydiary.R
import me.blog.korn123.easydiary.models.PhotoUriDto
import org.apache.commons.io.FileUtils
import org.apache.commons.io.IOUtils
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException


/**
 * Created by CHO HANJOONG on 2017-03-16.
 */

object CommonUtils {
    fun dpToPixelFloatValue(context: Context, dp: Int): Float {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp.toFloat(), context.resources.displayMetrics)
    }

    fun dpToPixel(context: Context, dp: Int, policy: Int = 0): Int {
        val px = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp.toFloat(), context.resources.displayMetrics)
        var pixel = 0
        when (policy) {
            0 -> pixel = px.toInt()
            1 -> pixel = Math.round(px)
        }
        return pixel
    }

    fun uriToPath(contentResolver: ContentResolver, uri: Uri): String? {
        var path: String? = null
        val cursor = contentResolver.query(uri, null, null, null, null)
        cursor!!.moveToNext()
        val columnIndex = cursor.getColumnIndex("_data")
        if (columnIndex > 0) {
            path = cursor.getString(columnIndex)
        } else {
            path = uri.toString()
        }
        cursor.close()
        return path
    }

    fun uriToFile(context: Context, uri: Uri, photoPath: String): Boolean {
        var result = false
        try {
            val tempFile = File.createTempFile("TEMP_PHOTO", "AAF").apply { deleteOnExit() }
            val inputStream = context.contentResolver.openInputStream(uri)
            val outputStream = FileOutputStream(tempFile)
            IOUtils.copy(inputStream, outputStream)
            IOUtils.closeQuietly(inputStream)
            IOUtils.closeQuietly(outputStream)

            val compressedFile = Compressor(context).setQuality(70).compressToFile(tempFile)
            FileUtils.copyFile(compressedFile, File(photoPath))
            result = true
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return result
    }
    
    fun photoUriToDownSamplingBitmap(
            context: Context,
            photoUriDto: PhotoUriDto,
            requiredSize: Int = 50,
            fixedWidth: Int = 45,
            fixedHeight: Int = 45
    ): Bitmap = try {
            when (photoUriDto.isContentUri()) {
                true -> {
                    BitmapUtils.decodeUri(context, Uri.parse(photoUriDto.photoUri), CommonUtils.dpToPixel(context, requiredSize, 1), CommonUtils.dpToPixel(context, fixedWidth, 1), CommonUtils.dpToPixel(context, fixedHeight, 1))
                }
                false -> {
                    BitmapUtils.decodeFile(context, photoUriDto.getFilePath(), CommonUtils.dpToPixel(context, requiredSize, 1), CommonUtils.dpToPixel(context, fixedWidth, 1), CommonUtils.dpToPixel(context, fixedHeight, 1))
                }
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
            BitmapFactory.decodeResource(context.resources, R.drawable.question_shield)
        } catch (se: SecurityException) {
            se.printStackTrace()
            BitmapFactory.decodeResource(context.resources, R.drawable.question_shield)
        }
    

    fun photoUriToBitmap(context: Context, photoUriDto: PhotoUriDto): Bitmap? {
        val bitmap: Bitmap? = try {
            when (photoUriDto.isContentUri()) {
                true -> {
                    BitmapFactory.decodeStream(context.contentResolver.openInputStream(Uri.parse(photoUriDto.photoUri)))
                }
                false -> {
                    BitmapFactory.decodeFile(photoUriDto.getFilePath())
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
           null
        }
        return bitmap
    }

    fun getDefaultDisplay(activity: Activity): Point {
        val display = activity.windowManager.defaultDisplay
        val size = Point()
        display.getSize(size)
        return size
    }
}
